streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -95,6 +97,7 @@
     private static final String JMX_PREFIX = "kafka.streams";
     public static final int DEFAULT_CLOSE_TIMEOUT = 0;
     private final StreamThread[] threads;
+    private final Map<Long, StreamThread.State> threadState;
```

Comment:
Actually it looks like it is `synchronized(stateListener)` in `StreamThread`. But even that seems a bit wrong. I think the synchronization would be better to be done in this class rather than elsewhere